### PR TITLE
Superposition: Add tasks from the contest

### PR DESCRIPTION
**Pull request by tcNickolas**
These tasks have been featured in Microsoft Q# Coding Contest - Winter 2019 (https://codeforces.com/contest/1116)

### CODE REVIEW

File: Superposition/ReferenceImplementation.qs [hunk 1]
```diff
@@ -178,7 +178,46 @@ namespace Quantum.Kata.Superposition {
...
+            // Split the state sqrt(2) |0⟩ ⊗ |0⟩ into |00⟩ + |01⟩
+            (ControlledOnInt(0, H))([qs[0]], qs[1]);
+        }
+        
+        adjoint auto;
+    }
+
+    // Alternative solution, based on post-selection
+    operation ThreeStates_TwoQubits_Postselection (qs : Qubit[]) : Unit {
+        using (ancilla = Qubit()) {
```
Comment (cgranade) [inline Superposition/ReferenceImplementation.qs]: In general, I suggest "auxillary" in preference to "ancilla."

File: Superposition/ReferenceImplementation.qs [hunk 2]
```diff
@@ -262,8 +302,55 @@ namespace Quantum.Kata.Superposition {
...
+    // Inputs:
+    //      1) N qubits in |0...0⟩ state
+    //      2) four bit string represented as Bool[][] bits
+    //         bits is an array of size 4 x N which describes the bit strings as follows:
+    //         bits[i] describes the i-th bit string and has N elements;
+    //         bit values false and true correspond to |0⟩ and |1⟩ states.
+    //
+    // Goal: create an equal superposition of the four basis states given by the bit strings.
+    //
+    // Example: for N = 2 and bits = [[false, true], [true, false], [false, false], [true, true]]
```
Comment (cgranade) [inline Superposition/ReferenceImplementation.qs]: Given that there are only four computational basis states for two qubits, perhaps a three-qubit example would make it more clear what this task covers?

File: Superposition/ReferenceImplementation.qs [hunk 3]
```diff
@@ -262,8 +302,55 @@ namespace Quantum.Kata.Superposition {
...
+    //          the state you need to prepare is (|01⟩ + |10⟩ + |00⟩ + |11⟩) / 2.
+    operation FourBitstringSuperposition_Reference (qs : Qubit[], bits : Bool[][]) : Unit {
+        body (...) {
+            let N = Length(qs);
+            
+            using (anc = Qubit[2]) {
```
Comment (cgranade) [inline Superposition/ReferenceImplementation.qs]: See above about "ancilla" versus "auxillary."

File: Superposition/Tasks.qs [hunk 4]
```diff
@@ -163,25 +172,43 @@ namespace Quantum.Kata.Superposition {
...
+    // Inputs:
+    //      1) N qubits in |0...0⟩ state
+    //      2) four bit string represented as Bool[][] bits
+    //         bits is an array of size 4 x N which describes the bit strings as follows:
+    //         bits[i] describes the i-th bit string and has N elements.
+    //         All four bit strings will be distinct.
+    //
+    // Goal: create an equal superposition of the four basis states given by the bit strings.
+    //
+    // Example: for N = 2 and bits = [[false, true], [true, false], [false, false], [true, true]]
```
Comment (cgranade) [inline Superposition/Tasks.qs]: Same comment as on the reference solution.

File: Superposition/ReferenceImplementation.qs (hunk 1 above)
Comment (tcNickolas) [reply Superposition/ReferenceImplementation.qs]: We use both terms interchangeably across our code base, so it's fine.

**Pull request merged**